distribution/lib/Standard/Base/0.0.0-dev/src/System/File.enso

```diff
@@ -832,6 +832,12 @@ type File
     to_display_text : Text
     to_display_text self = self.to_text
 
+    ## PRIVATE
+    copy_builtin self target options = @Builtin_Method "File.copy_builtin"
```

Comment:
Seeing all these (undesirable) `@Builtin_Method`s changes, I wonder how are they related to _type inference_?

Reply:
There are no _changes_ to builtin methods structure.

I'm only reproducing in Enso code the structure that **has already been there** in the builtins. That is - we already were having calls to `File.copy_builtin` or `Comparable.less_than_builtin` in the code, they were just lacking the related definitions.

If that structure is _undesirable_, we should restructure the builtins, but that is not really in the scope of the PR. But if you will not allow me to merge it without that, I will probably be forced to do this _unrelated_ change.


> I wonder how are they related to _type inference_?

I've added checking of method calls. What methods are available on an object was taken from the type definitions. If a builtin method was called e.g. `File.copy_builtin` but there was no `copy_builtin` defined on `File`, my type inference was giving warnings about calling nonexstent methods. I don't think builtins shall be a special case (apart from a naming convention I cannot easily tell if a method is a builtin or doesn't exist). Most of our builtins already had such 'shadow definitions', there was just a few that were lacking and I've added these missing ones. See for example, `File.input_stream_builtin` - it's been there for a long time. I think we should do a proper refactor of builtins to move them into private modules. But that's the #6282 ticket.

Reply:
On the other hand, these changes are not absolutely **required** for the type inference changes. They are there to make the tests pass, but I can `@Ignore` a few tests and revert these changes so that we can handle them separately. Maybe I should have done them as a separate PR like with #11422

Reply:
I like these kind of *shadow* definitions of builtin methods. They don't (at least they should not) introduce any changes (that is, they don't introduce any new builtins) and they make it easy to see which builtin methods are defined on a type. It may be worth revisiting [BuiltinMethod.autoRegister](https://github.com/enso-org/enso/blob/c2c6712b77cf8fc7aba8a478e6169ee1dceb23ce/lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/BuiltinMethod.java#L41) annotation argument.

See [my other comment](https://github.com/enso-org/enso/pull/11399#discussion_r1827559455)

Reply:
> If a builtin method was called e.g. `File.copy_builtin` but there was no `copy_builtin` defined on `File`, my type inference was giving warnings about calling nonexstent methods. 

I see. That's indeed related.

> I don't think builtins shall be a special case 

Builtin methods are _special case_! They are not behaving as normal methods. 

> we should do a proper refactor of builtins to move them into private modules. ... that's the #6282 ticket.

> ... shadow definitions of builtin methods ... They don't introduce any changes ...

OK.

> worth revisiting [BuiltinMethod.autoRegister](https://github.com/enso-org/enso/blob/c2c6712b77cf8fc7aba8a478e6169ee1dceb23ce/lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/BuiltinMethod.java#L41) annotation argument.

- having two places to specify the same thing feel weird
- introduced by @hubertp in #3791

> Builtins...carry a `autoRegister` property to determine if a builtin method should be automatically registered with the type.

What's the benefit of _registering a builtin automatically with the type_, @hubertp?